lib/matplotlib/axes/_axes.py

```diff
@@ -5259,7 +5259,27 @@ def _pcolorargs(funcname, *args, allmatch=False):
                 raise TypeError('Dimensions of C %s are incompatible with'
                                 ' X (%d) and/or Y (%d); see help(%s)' % (
                                     C.shape, Nx, Ny, funcname))
-            C = C[:Ny - 1, :Nx - 1]
+
+            if dropdata:
+                 C = C[:Ny - 1, :Nx - 1]
+            else:
+                def _interp_grid(X):
```

Comment:
```suggestion
                def _interp_grid(X):
                    # helper for below
                    dX = np.diff(X, axis=1)/2.
                    if np.allclose(dX[:, :], dX[0, [0]]):
                        X = np.hstack((X[:, [0]] - dX[:, [0]],
                                       X[:, :-1] + dX,
                                       X[:, [-1]] + dX[:, [-1]])
                                      )
                    else:
                        X_ratio = np.sqrt(X[:, 1:]/X[:, 0:-1])
                        if np.allclose(X_ratio[:, :], X_ratio[0, [0]]):
                            X = np.hstack((X[:, [0]]/X_ratio[:, [0]],
                                           X[:, :-1]*X_ratio,
                                           X[:, [-1]]*X_ratio[:, [-1]])
                                          )
                        else:
                            pass
                            # Warn the user that the data is neither linearly
                            # nor logarithmically spaced.
                    return X                   
```
The changes I am suggesting take care of the case when X and/or Y are logarithmically spaced.
(e.g. visualizing a time series of fractional octave band sound pressure level data)

```python
import numpy as np
import matplotlib
import matplotlib.pyplot as plt

time = np.array([0, 1, 2, 3])
freq = np.array([500, 1000, 2000, 4000])
time_mesh, freq_mesh = np.meshgrid(time, freq)
C = np.arange(16).reshape(4,4)

fig, ax = plt.subplots()
im = ax.pcolormesh(time_mesh, freq_mesh, C, dropdata=False)
cbar = fig.colorbar(im)
cbar.set_label('Octave band SPL [dB$_{ref: 20 \mu Pa}$]')

ax.set_xticks(time)
ax.set_yscale('log')
ax.set_yticks(freq)
ax.set_yticklabels(freq)
ax.minorticks_off()
ax.set_xlabel('Time [s]')
ax.set_ylabel('Octave band center frequency  [Hz]')

plt.show()
```
![figure_1](https://user-images.githubusercontent.com/15175620/47270666-40b31680-d53d-11e8-8780-38c7ae6abb9a.png)

Reply:
I like this logarithmic addition! The only thing I'd add is that you can still use the method and calculate dX's even if the points aren't evenly spaced, so the first if statement shouldn't be a requirement. It could just be the default after checking for the ratio.

Reply:
I don't think this will apply generally enough to be a good approach.  If someone needs precision in the cell edges, they should calculate them themselves as they need, not count on matplotlib to do it.  This is just a rough fix which is not quite as rough as just dropping the data.

Reply:
Yeah, there are a whole bunch of cases where this will fail/error out (`x = [10, 9, 8]`) that we could check for, but I'm not convinced its worth the extra code.

Reply:
I guess a possibility would be to do the grid interpolation in screen space?  Something like... (assuming dropdata=False and the shapes are such that grid interp is necessary)
- convert the x values using the current scale
- if the scaled values are uniformly spaced, do the interp, convert back to data space, use the interp'd grid.
- if not (and an interpolation is necessary), emit a warning (so that users will know to call set_xscale("log") before passing their log-scaled x values
(very rough sketch)

Reply:
That seems way too fancy and hard to explain to the user.  And I'm not even convinced it makes sense for curvilinear grids etc.  Lets keep this simple and easy to explain.  If the user wants something fancier they can calculate their own edges...

Reply:
> I guess a possibility would be to do the grid interpolation in screen space? Something like... (assuming dropdata=False and the shapes are such that grid interp is necessary)
> 
>     * convert the x values using the current scale
> 
>     * if the scaled values are uniformly spaced, do the interp, convert back to data space, use the interp'd grid.
> 
>     * if not (and an interpolation is necessary), emit a warning (so that users will know to call set_xscale("log") before passing their log-scaled x values

This process could be converted into a helper function which could be showed off in a gallery example/matplotlib blog. If the OP is fine with it, I am interested in working on a PR demonstrating the process outlined here. Does this sound a like a good idea?